pkg/dockerregistry/server/errorblobstore.go

```diff
@@ -23,44 +23,57 @@ func (r *errorBlobStore) Stat(ctx context.Context, dgst digest.Digest) (distribu
 	if err := r.repo.checkPendingErrors(ctx); err != nil {
 		return distribution.Descriptor{}, err
 	}
-	return r.store.Stat(ctx, dgst)
+	return r.store.Stat(WithRepository(ctx, r.repo), dgst)
 }
 
 func (r *errorBlobStore) Get(ctx context.Context, dgst digest.Digest) ([]byte, error) {
 	if err := r.repo.checkPendingErrors(ctx); err != nil {
 		return nil, err
 	}
-	return r.store.Get(ctx, dgst)
+	return r.store.Get(WithRepository(ctx, r.repo), dgst)
 }
 
 func (r *errorBlobStore) Open(ctx context.Context, dgst digest.Digest) (distribution.ReadSeekCloser, error) {
 	if err := r.repo.checkPendingErrors(ctx); err != nil {
 		return nil, err
 	}
-	return r.store.Open(ctx, dgst)
+	return r.store.Open(WithRepository(ctx, r.repo), dgst)
 }
 
 func (r *errorBlobStore) Put(ctx context.Context, mediaType string, p []byte) (distribution.Descriptor, error) {
 	if err := r.repo.checkPendingErrors(ctx); err != nil {
 		return distribution.Descriptor{}, err
 	}
-	return r.store.Put(ctx, mediaType, p)
+	return r.store.Put(WithRepository(ctx, r.repo), mediaType, p)
 }
 
 func (r *errorBlobStore) Create(ctx context.Context, options ...distribution.BlobCreateOption) (distribution.BlobWriter, error) {
+	var desc distribution.Descriptor
+
 	if err := r.repo.checkPendingErrors(ctx); err != nil {
 		return nil, err
 	}
 
+	ctx = WithRepository(ctx, r.repo)
+
 	opts, err := effectiveCreateOptions(options)
 	if err != nil {
 		return nil, err
 	}
-	if err := checkPendingCrossMountErrors(ctx, opts); err != nil {
-		context.GetLogger(r.repo.ctx).Debugf("disabling cross-mount because of pending error: %v", err)
+	err = checkPendingCrossMountErrors(ctx, opts)
+	if err == nil && opts.Mount.ShouldMount {
+		desc, err = statSourceRepository(ctx, opts.Mount.From, opts.Mount.From.Digest())
+	}
+
+	if err != nil {
+		context.GetLogger(ctx).Infof("disabling cross-repo mount because of an error: %v", err)
+		options = append(options, guardCreateOptions{DisableCrossMount: true})
+	} else if !opts.Mount.ShouldMount {
+		context.GetLogger(ctx).Infof("ensuring cross-repo is disabled: %v", err)
 		options = append(options, guardCreateOptions{DisableCrossMount: true})
 	} else {
-		options = append(options, guardCreateOptions{})
+		context.GetLogger(ctx).Debugf("attempting cross-repo mount")
+		options = append(options, statCrossMountCreateOptions{desc: desc})
```

Comment:
should we actually do the `statSourceRepository()` call inside `statCrossMountCreateOptions#Apply`?